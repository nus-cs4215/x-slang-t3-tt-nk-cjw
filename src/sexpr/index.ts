export * from './sexpr';
export * from './jsonsexpr';
