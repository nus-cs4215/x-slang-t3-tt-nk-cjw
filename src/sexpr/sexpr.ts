export interface SAtom {
  _type: 'SAtom';
  val: string;
}

export interface SNumber {
  _type: 'SNumber';
  val: number;
}

export interface SBoolean {
  _type: 'SBoolean';
  val: boolean;
}

export interface SNil {
  _type: 'SNil';
}

enum SListVariants {
  PAIR,
}

interface SListBase {
  _type: 'SList';
  _variant: SListVariants;
}

interface SListPair extends SListBase {
  _variant: SListVariants.PAIR;

  x: SExpr;
  y: SExpr;
}

export type SList = SListPair;

export type SExpr = SAtom | SNumber | SBoolean | SNil | SList;

export const satom = (val: string): SAtom => ({ _type: 'SAtom', val });
export const snumber = (val: number): SNumber => ({ _type: 'SNumber', val });
export const sboolean = (val: boolean): SBoolean => ({
  _type: 'SBoolean',
  val,
});

export const snil = (): SNil => ({ _type: 'SNil' });
export const scons = (x: SExpr, y: SExpr): SList => ({
  _type: 'SList',
  _variant: SListVariants.PAIR,
  x,
  y,
});

export function slist(xs: [SExpr, ...SExpr[]], tail: SExpr): SList;
export function slist(xs: SExpr[], tail: SExpr): SExpr;
export function slist(xs: SExpr[], tail: SExpr): SExpr {
  let p: SExpr = tail;
  for (let i = xs.length - 1; i >= 0; i--) {
    p = scons(xs[i], p);
  }
  return p;
}

export const car = (p: SList): SExpr => p.x;
export const cdr = (p: SList): SExpr => p.y;

export function* sconslist_iterator(p: SList): Iterable<SExpr | '.'> & Iterator<SExpr | '.'> {
  let e: SExpr;
  for (e = p; e._type === 'SList'; e = cdr(e)) {
    yield car(e);
  }
  if (e._type !== 'SNil') {
    yield '.';
    yield e;
  }
}

export function equals<E2 extends SExpr>(e1: SExpr, e2: E2): e1 is E2;
export function equals<E1 extends SExpr>(e1: E1, e2: SExpr): e2 is E1;
export function equals(e1: SExpr, e2: SExpr): boolean {
  while (true) {
    if (e1._type === 'SAtom' && e2._type === 'SAtom') {
      return e1.val === e2.val;
    } else if (e1._type === 'SNumber' && e2._type === 'SNumber') {
      return e1.val === e2.val;
    } else if (e1._type === 'SBoolean' && e2._type === 'SBoolean') {
      return e1.val === e2.val;
    } else if (e1._type === 'SNil' && e2._type === 'SNil') {
      return true;
    } else if (e1._type === 'SList' && e2._type === 'SList') {
      if (!equals(car(e1), car(e2))) {
        return false;
      }
      e1 = cdr(e1);
      e2 = cdr(e2);
    } else {
      return false;
    }
  }
}

export type JsonSExpr = JsonSExpr[] | string | number | boolean;
