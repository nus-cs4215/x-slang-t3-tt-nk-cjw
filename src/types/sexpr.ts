export type SAtom = string;
export type SNil = [];
export type SExpr = SAtom | [SExpr[], SExpr] | SNil;
