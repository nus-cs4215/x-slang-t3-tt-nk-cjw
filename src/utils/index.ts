export * from './result';
export * from './location';
export * from './formatters';
export * from './types';
