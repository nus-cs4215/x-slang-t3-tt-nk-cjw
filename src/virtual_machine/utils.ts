import { CompiledProgram } from './compiler';
import { get_opcode_names, get_opcode_paramCounts } from './opcodes';

export const prettify_compiled_program = (compiledProgram: CompiledProgram): string[] => {
  const opcodeNames = get_opcode_names();
  const paramCounts = get_opcode_paramCounts();
  const prettified: string[] = [];
  for (let i = 0; i < compiledProgram.length; i++) {
    const opcode = compiledProgram[i];
    if (Array.isArray(opcode)) {
      prettified.push('EARLY TERMINATION. Compiled Program not flattened');
      return prettified;
    }

    prettified.push(opcodeNames[opcode]); // push opcodes

    // push the n parameters behind it
    let paramCount = paramCounts[opcode];
    while (paramCount > 0) {
      prettified.push(compiledProgram[++i].toString());
      paramCount--;
    }
  }
  return prettified;
};
