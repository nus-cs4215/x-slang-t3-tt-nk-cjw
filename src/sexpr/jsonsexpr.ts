import { sbox, SExprT } from './sexpr';
import { val, car, cdr } from './sexpr';
import { ssymbol, snumber, sboolean, snil, slist } from './sexpr';
import { is_symbol, is_number, is_boolean, is_nil, is_list, is_boxed } from './sexpr';

type JsonBoxed<T> = T extends never ? never : { boxed: T };

const jbox = <T>(boxed: T): JsonBoxed<T> => ({ boxed } as JsonBoxed<T>);

export type JsonSExpr<T> = JsonSExpr<T>[] | string | number | boolean | JsonBoxed<T>;

export function jsonPrint<T>(e: SExprT<T>): JsonSExpr<T> {
  if (is_symbol(e) || is_number(e) || is_boolean(e)) {
    return val(e);
  } else if (is_nil(e)) {
    return [];
  } else if (is_boxed(e)) {
    return jbox(e.val);
  } else {
    // if (e._type === STypes.List) {
    const xs = [];
    xs.push(jsonPrint(car(e)));
    e = cdr(e);
    for (; is_list(e); e = cdr(e)) {
      xs.push(jsonPrint(car(e)));
    }
    if (!is_nil(e)) {
      xs.push('.');
      xs.push(jsonPrint(e));
    }
    return xs;
  }
}

export function jsonRead<T>(j: JsonSExpr<T>): SExprT<T> {
  if (typeof j === 'string') {
    return ssymbol(j);
  } else if (typeof j === 'number') {
    return snumber(j);
  } else if (typeof j === 'boolean') {
    return sboolean(j);
  } else if (Array.isArray(j)) {
    if (j.length === 0) {
      return snil();
    }
    if (j.length >= 3 && j[j.length - 2] === '.') {
      const tail = jsonRead(j.pop()!);
      j.pop();
      return slist(j.map(jsonRead), tail);
    }
  } else {
    // j is object, so it's the boxed type
    return sbox(j.boxed);
  }
  return slist(j.map(jsonRead), snil());
}
