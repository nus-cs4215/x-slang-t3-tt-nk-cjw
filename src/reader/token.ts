import { Position, Location } from '../utils/location';

export interface Base {
  type: string;
  contents: string;
  loc: Location;
}

export interface LPar extends Base {
  type: 'LPar';
}

export function lpar(contents: string, loc: Location): LPar {
  return { type: 'LPar', contents, loc };
}

export interface RPar extends Base {
  type: 'RPar';
}

export function rpar(contents: string, loc: Location): RPar {
  return { type: 'RPar', contents, loc };
}

export interface Quote extends Base {
  type: 'Quote';
}

export function quote(contents: string, loc: Location): Quote {
  return { type: 'Quote', contents, loc };
}

export interface Dot extends Base {
  type: 'Dot';
}

export function dot(contents: string, loc: Location): Dot {
  return { type: 'Dot', contents, loc };
}

export interface Atom extends Base {
  type: 'Atom';
}

export function atom(contents: string, loc: Location): Atom {
  return { type: 'Atom', contents, loc };
}

export interface Num extends Base {
  type: 'Num';
}

export function num(contents: string, loc: Location): Num {
  return { type: 'Num', contents, loc };
}

export interface Bool extends Base {
  type: 'Bool';
}

export function bool(contents: string, loc: Location): Bool {
  return { type: 'Bool', contents, loc };
}

export interface Invalid extends Base {
  type: 'Invalid';
}

export function invalid(contents: string, loc: Location): Invalid {
  return { type: 'Invalid', contents, loc };
}

export interface EOF extends Base {
  type: 'EOF';
  contents: '';
}

export function eof(loc: Location): EOF {
  return { type: 'EOF', contents: '', loc };
}

export type Token = LPar | RPar | Quote | Dot | Atom | Num | Bool | Invalid | EOF;

export function par_match(lpar: LPar, rpar: RPar): boolean {
  return (
    (lpar.contents === '(' && rpar.contents === ')') ||
    (lpar.contents === '[' && rpar.contents === ']') ||
    (lpar.contents === '{' && rpar.contents === '}')
  );
}

function substring_to_token(s: string, loc: Location): Token {
  // Try to interpret as single dot
  if (s.length === 1 && s[0] === '.') {
    return dot('.', loc);
  }

  // Try to interpret as boolean
  if (s[0] === '#') {
    if (s.length !== 2) {
      return invalid(s, loc);
    }
    if (s[1] === 't') {
      return bool('#t', loc);
    } else if (s[1] === 'f') {
      return bool('#f', loc);
    }

    return invalid(s, loc);
  }

  // try to interpret as number
  if (Number.isFinite(Number(s))) {
    return num(s, loc);
  }

  // I guess it's an atom, :P
  return atom(s, loc);
}

const ws_regex = /\s/;
const delim_regex = /\s|[()[\]{}",'`;]/;

export function* tokenize(s: string): Iterable<Token> & Iterator<Token> {
  let i = 0;
  let line = 1;
  let column = 0;

  function inc(): void {
    if (s[i] === '\n') {
      line++;
      column = 0;
      i++;
    } else {
      column++;
      i++;
    }
  }

  function pos(): Position {
    return {
      character: i,
      line,
      column,
    };
  }

  while (true) {
    if (i === s.length) {
      const curpos = pos();
      yield eof({ start: curpos, end: curpos });
      return;
    }

    // strip whitespace
    while (ws_regex.test(s[i])) {
      inc();
      if (i === s.length) {
        const curpos = pos();
        yield eof({ start: curpos, end: curpos });
        return;
      }
    }

    const start = pos();
    // Read delimiter character
    switch (s[i]) {
      case '(':
      case '[':
      case '{': {
        const contents = s[i];
        inc();
        const end = pos();
        yield lpar(contents, { start, end });
        continue;
      }
      case ')':
      case ']':
      case '}': {
        const contents = s[i];
        inc();
        const end = pos();
        yield rpar(contents, { start, end });
        continue;
      }
      case "'": {
        const contents = s[i];
        inc();
        const end = pos();
        yield quote(contents, { start, end });
        continue;
      }
      case ';': {
        // line comment, skip to end of line
        while (s[i] !== '\n') {
          inc();
          if (i === s.length) {
            const curpos = pos();
            yield eof({ start: curpos, end: curpos });
            return;
          }
        }
        continue;
      }
    }

    // Try to read a symbol
    const j = i;
    while (!delim_regex.test(s[i])) {
      inc();
      if (i === s.length) {
        break;
      }
    }

    const contents = s.slice(j, i);
    const end = pos();
    yield substring_to_token(contents, { start, end });
  }

  const curpos = pos();
  yield eof({ start: curpos, end: curpos });
}
