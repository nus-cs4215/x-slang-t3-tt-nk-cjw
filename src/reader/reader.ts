import { Result, isGoodResult, ok, err, then } from '../utils';
import { SExpr, satom, snumber, sboolean, snil, scons, slist } from '../sexpr';
import * as Tok from './token';
import { Token, tokenize, par_match } from './token';
import { Location, merge_loc, format_loc, highlight_loc } from '../utils/location';

export interface ReadErr {
  message: string;
  loc: Location;
}

export type PartialReadResult = Result<[SExpr, Location, number], ReadErr>;

export type ReadResult = Result<SExpr, ReadErr>;

export function formatReadErr(err: ReadErr, s: string) {
  return `Read error at ${format_loc(err.loc)}: ${err.message}
${highlight_loc(err.loc, s, '  ')}
`;
}

function token_to_sexpr(tok: Tok.Atom | Tok.Num | Tok.Bool): [SExpr, Location] {
  switch (tok.type) {
    case 'Atom': {
      return [satom(tok.contents), tok.loc];
    }
    case 'Num': {
      return [snumber(Number.parseFloat(tok.contents)), tok.loc];
    }
    case 'Bool': {
      return [sboolean(tok.contents === '#t'), tok.loc];
    }
  }
}

function readList(firsttok: Tok.LPar, tokens: Token[], i: number): PartialReadResult {
  const data1: SExpr[] = [];
  let firstdot: Tok.Dot | undefined = undefined;
  const data2: SExpr[] = [];
  let seconddot: Tok.Dot | undefined = undefined;
  const data3: SExpr[] = [];

  // Stage 1, no . seen yet
  while (true) {
    const curtok = tokens[i];

    // Ending at stage 1, no . in this list
    if (curtok.type === 'RPar') {
      i++;
      // Check that the parens were matched
      if (!par_match(firsttok, curtok)) {
        return err({
          message: 'Mismatched parentheses',
          loc: merge_loc(firsttok.loc, curtok.loc),
        });
      }

      if (data1.length === 0) {
        return ok([snil(), merge_loc(firsttok.loc, curtok.loc), i]);
      } else if (data1.length === 1) {
        return ok([scons(data1[0], snil()), merge_loc(firsttok.loc, curtok.loc), i]);
      } else {
        return ok([slist(data1, snil()), merge_loc(firsttok.loc, curtok.loc), i]);
      }
    }

    if (curtok.type === 'Dot') {
      if (data1.length === 0) {
        // Nothing before the . is illegal
        return err({
          message: 'Unexpected dot, missing LHS of cons literal or infix list',
          loc: merge_loc(firsttok.loc, curtok.loc),
        });
      }

      // Move to next stage
      firstdot = curtok;
      i++;
      break;
    }

    const res = readOneDatum(tokens, i);
    if (isGoodResult(res)) {
      let datum: SExpr;
      let _loc: Location | undefined = undefined;
      _loc; // unused
      [datum, _loc, i] = res.v;
      data1.push(datum);
    } else {
      return res;
    }
  }

  // Stage 2, one . seen
  while (true) {
    const curtok = tokens[i];

    // Ending at stage 2, one . in this list
    if (curtok.type === 'RPar') {
      if (data2.length === 0) {
        // Nothing coming after the . is illegal
        return err({
          message: 'Unexpected parenthesis, missing RHS of cons literal',
          loc: curtok.loc,
        });
      } else if (data2.length !== 1) {
        // More than 1 datum after the . is also illegal
        return err({
          message: 'Too much data on the RHS of cons literal',
          loc: merge_loc(firstdot.loc, curtok.loc),
        });
      }

      i++;
      // Check that the parens were matched
      if (!par_match(firsttok, curtok)) {
        return err({
          message: 'Mismatched parentheses',
          loc: merge_loc(firsttok.loc, curtok.loc),
        });
      }

      if (data1.length === 1) {
        return ok([scons(data1[0], data2[0]), merge_loc(firsttok.loc, curtok.loc), i]);
      } else {
        return ok([slist(data1, data2[0]), merge_loc(firsttok.loc, curtok.loc), i]);
      }
    }

    if (curtok.type === 'Dot') {
      if (data2.length === 0) {
        // Nothing between the . is illegal
        return err({
          message: 'Unexpected dot, missing data between dots of infix list',
          loc: merge_loc(firstdot.loc, curtok.loc),
        });
      }

      if (data2.length !== 1) {
        // Too many data between the . is illegal too
        return err({
          message: 'Unexpected dot, too many data between dots of infix list',
          loc: merge_loc(firstdot.loc, curtok.loc),
        });
      }

      // Move to next stage
      seconddot = curtok;
      i++;
      break;
    }

    const res = readOneDatum(tokens, i);
    if (isGoodResult(res)) {
      let datum: SExpr;
      let _loc: Location | undefined = undefined;
      _loc; // unused
      [datum, _loc, i] = res.v;
      data2.push(datum);
    } else {
      return res;
    }
  }

  // Stage 3, two . seen
  while (true) {
    const curtok = tokens[i];

    // Ending at stage 3, two . in this list
    if (curtok.type === 'RPar') {
      if (data3.length === 0) {
        // Nothing coming after the . is illegal
        return err({
          message: 'Unexpected parenthesis, missing RHS of infix list',
          loc: merge_loc(seconddot.loc, curtok.loc),
        });
      }

      i++;
      // Check that the parens were matched
      if (!par_match(firsttok, curtok)) {
        return err({
          message: 'Mismatched parentheses',
          loc: merge_loc(firsttok.loc, curtok.loc),
        });
      }

      const data = [...data2, ...data1, ...data3];
      return ok([slist(data, snil()), merge_loc(firsttok.loc, curtok.loc), i]);
    }

    if (curtok.type === 'Dot') {
      // Too many dots
      return err({
        message: 'Too many dots in infix list',
        loc: curtok.loc,
      });
    }

    const res = readOneDatum(tokens, i);
    if (isGoodResult(res)) {
      let datum: SExpr;
      let _loc: Location | undefined = undefined;
      _loc; // unused
      [datum, _loc, i] = res.v;
      data3.push(datum);
    } else {
      return res;
    }
  }
}

export function readOneDatum(tokens: Token[], i: number = 0): PartialReadResult {
  const firsttok = tokens[i];

  if (firsttok.type === 'EOF') {
    return err({ message: 'Unexpected EOF', loc: firsttok.loc });
  }

  if (firsttok.type === 'Invalid') {
    return err({ message: 'Invalid token', loc: firsttok.loc });
  }

  if (firsttok.type === 'RPar') {
    return err({ message: 'Unexpected parenthesis', loc: firsttok.loc });
  }

  if (firsttok.type === 'Dot') {
    return err({ message: 'Unexpected dot', loc: firsttok.loc });
  }

  // Parse list
  if (firsttok.type === 'LPar') {
    return readList(firsttok, tokens, i + 1);
  }

  // Parse quoted
  if (firsttok.type === 'Quote') {
    return then(readOneDatum(tokens, i + 1), ([e, loc, i]) =>
      ok([slist([satom('quote'), e], snil()), merge_loc(firsttok.loc, loc), i])
    );
  }

  // Everything else is just itself
  return ok([...token_to_sexpr(firsttok), i + 1]);
}

export function read(s: string): ReadResult {
  return then(readOneDatum([...tokenize(s)]), (r: [SExpr, Location, number]) => ok(r[0]));
}
