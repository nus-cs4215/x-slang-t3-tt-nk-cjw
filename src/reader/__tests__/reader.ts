import { getOk, getErr } from '../../utils';
import { equals, jsonPrint } from '../../sexpr';
import { read, formatReadErr } from '../reader';

function expectReadAsJson(s: string) {
  return expect(jsonPrint(getOk(read(s))));
}

describe('valid read tests', () => {
  test('basic parens', () => {
    expectReadAsJson('()').toMatchInlineSnapshot(`Array []`);
    expectReadAsJson('[]').toMatchInlineSnapshot(`Array []`);
    expectReadAsJson('{}').toMatchInlineSnapshot(`Array []`);
  });

  test('basic symbols', () => {
    expectReadAsJson('abc').toMatchInlineSnapshot(`"abc"`);
    expectReadAsJson('abc->def').toMatchInlineSnapshot(`"abc->def"`);
  });

  test('basic numbers', () => {
    expectReadAsJson('123').toMatchInlineSnapshot(`123`);
    expectReadAsJson('123.5').toMatchInlineSnapshot(`123.5`);
  });

  test('basic booleans', () => {
    expectReadAsJson('#t').toMatchInlineSnapshot(`true`);
    expectReadAsJson('#f').toMatchInlineSnapshot(`false`);
  });

  test('basic lists', () => {
    expectReadAsJson('()').toMatchInlineSnapshot(`Array []`);
    expectReadAsJson('(abc)').toMatchInlineSnapshot(`
      Array [
        "abc",
      ]
    `);
    expectReadAsJson('(abc 123)').toMatchInlineSnapshot(`
      Array [
        "abc",
        123,
      ]
    `);
    expectReadAsJson('(abc 123 #t)').toMatchInlineSnapshot(`
      Array [
        "abc",
        123,
        true,
      ]
    `);
  });

  test('basic cons', () => {
    expectReadAsJson('(a . b)').toMatchInlineSnapshot(`
      Array [
        "a",
        ".",
        "b",
      ]
    `);
    expectReadAsJson('(a b . c)').toMatchInlineSnapshot(`
      Array [
        "a",
        "b",
        ".",
        "c",
      ]
    `);
    expect(equals(getOk(read('(a . ())')), getOk(read('(a)')))).toBe(true);
  });

  test('basic infix', () => {
    expect(equals(getOk(read('(a . b . c)')), getOk(read('(b a c)')))).toBe(true);
    expect(equals(getOk(read('(a b . c . d e)')), getOk(read('(c a b d e)')))).toBe(true);
  });

  test('basic quotes', () => {
    expectReadAsJson("'a").toMatchInlineSnapshot(`
      Array [
        "quote",
        "a",
      ]
    `);
    expectReadAsJson('`a').toMatchInlineSnapshot(`
      Array [
        "quasiquote",
        "a",
      ]
    `);
    expectReadAsJson(',a').toMatchInlineSnapshot(`
      Array [
        "unquote",
        "a",
      ]
    `);
    expectReadAsJson("'()").toMatchInlineSnapshot(`
      Array [
        "quote",
        Array [],
      ]
    `);
    expectReadAsJson('`(+ ,(+ 1 2) ,(+ 3 4))').toMatchInlineSnapshot(`
      Array [
        "quasiquote",
        Array [
          "+",
          Array [
            "unquote",
            Array [
              "+",
              1,
              2,
            ],
          ],
          Array [
            "unquote",
            Array [
              "+",
              3,
              4,
            ],
          ],
        ],
      ]
    `);
    expectReadAsJson("('abc)").toMatchInlineSnapshot(`
      Array [
        Array [
          "quote",
          "abc",
        ],
      ]
    `);
    expectReadAsJson("'(abc)").toMatchInlineSnapshot(`
      Array [
        "quote",
        Array [
          "abc",
        ],
      ]
    `);
    expectReadAsJson("'(abc . def)").toMatchInlineSnapshot(`
      Array [
        "quote",
        Array [
          "abc",
          ".",
          "def",
        ],
      ]
    `);
    expectReadAsJson("''(abc)").toMatchInlineSnapshot(`
      Array [
        "quote",
        Array [
          "quote",
          Array [
            "abc",
          ],
        ],
      ]
    `);
    expectReadAsJson("'('abc)").toMatchInlineSnapshot(`
      Array [
        "quote",
        Array [
          Array [
            "quote",
            "abc",
          ],
        ],
      ]
    `);
  });
});

function expectFormattedReadErr(s: string) {
  return expect(formatReadErr(getErr(read(s)), s));
}

describe('invalid read tests', () => {
  test('bad parens', () => {
    expectFormattedReadErr('(').toMatchInlineSnapshot(`
      "Read error at 1:1 to 1:1: Unexpected EOF
        1 | (
          |  ^

      "
    `);

    expectFormattedReadErr(')').toMatchInlineSnapshot(`
      "Read error at 1:0 to 1:1: Unexpected parenthesis
        1 | )
          | ^

      "
    `);

    expectFormattedReadErr('(]').toMatchInlineSnapshot(`
      "Read error at 1:0 to 1:2: Mismatched parentheses
        1 | (]
          | ^~

      "
    `);

    expectFormattedReadErr('([ abc def )]').toMatchInlineSnapshot(`
      "Read error at 1:1 to 1:12: Mismatched parentheses
        1 | ([ abc def )]
          |  ^~~~~~~~~~~

      "
    `);

    expectFormattedReadErr('([ abc\ndef )]').toMatchInlineSnapshot(`
      "Read error at 1:1 to 2:5: Mismatched parentheses
        1 | ([ abc
          |  ^~~~~
        2 | def )]
          | ~~~~~

      "
    `);

    expectFormattedReadErr('([ abc\ndef\nghi )]').toMatchInlineSnapshot(`
      "Read error at 1:1 to 3:5: Mismatched parentheses
        1   | ([ abc
            |  ^~~~~
        ... | ...
        3   | ghi )]
            | ~~~~~

      "
    `);
  });

  test('bad symbols', () => {
    expectFormattedReadErr('.').toMatchInlineSnapshot(`
      "Read error at 1:0 to 1:1: Unexpected dot
        1 | .
          | ^

      "
    `);
  });

  test('bad delims', () => {
    expectFormattedReadErr("'").toMatchInlineSnapshot(`
      "Read error at 1:1 to 1:1: Unexpected EOF
        1 | '
          |  ^

      "
    `);
  });

  test('bad cons', () => {
    expectFormattedReadErr('(a .)').toMatchInlineSnapshot(`
      "Read error at 1:4 to 1:5: Unexpected parenthesis, missing RHS of cons literal
        1 | (a .)
          |     ^

      "
    `);

    expectFormattedReadErr('(a . b c)').toMatchInlineSnapshot(`
      "Read error at 1:3 to 1:9: Too much data on the RHS of cons literal
        1 | (a . b c)
          |    ^~~~~~

      "
    `);

    expectFormattedReadErr('(. b)').toMatchInlineSnapshot(`
      "Read error at 1:0 to 1:2: Unexpected dot, missing LHS of cons literal or infix list
        1 | (. b)
          | ^~

      "
    `);
  });

  test('bad infix', () => {
    expectFormattedReadErr('(a . b .)').toMatchInlineSnapshot(`
      "Read error at 1:7 to 1:9: Unexpected parenthesis, missing RHS of infix list
        1 | (a . b .)
          |        ^~

      "
    `);

    expectFormattedReadErr('(a . . b)').toMatchInlineSnapshot(`
      "Read error at 1:3 to 1:6: Unexpected dot, missing data between dots of infix list
        1 | (a . . b)
          |    ^~~

      "
    `);

    expectFormattedReadErr('(. a . b)').toMatchInlineSnapshot(`
      "Read error at 1:0 to 1:2: Unexpected dot, missing LHS of cons literal or infix list
        1 | (. a . b)
          | ^~

      "
    `);

    expectFormattedReadErr('(a . b c . d)').toMatchInlineSnapshot(`
      "Read error at 1:3 to 1:10: Unexpected dot, too many data between dots of infix list
        1 | (a . b c . d)
          |    ^~~~~~~

      "
    `);
  });

  test('large line number', () => {
    expectFormattedReadErr('\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n)').toMatchInlineSnapshot(`
      "Read error at 19:0 to 19:1: Unexpected parenthesis
        19 | )
           | ^

      "
    `);

    expectFormattedReadErr('\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n(\n]').toMatchInlineSnapshot(`
      "Read error at 18:0 to 19:1: Mismatched parentheses
        18 | (
           | ^
        19 | ]
           | ~

      "
    `);

    expectFormattedReadErr('\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n(\n\n]').toMatchInlineSnapshot(`
      "Read error at 18:0 to 20:1: Mismatched parentheses
        18  | (
            | ^
        ... | ...
        20  | ]
            | ~

      "
    `);

    expectFormattedReadErr('(\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n]').toMatchInlineSnapshot(`
      "Read error at 1:0 to 19:1: Mismatched parentheses
        1   | (
            | ^
        ... | ...
        19  | ]
            | ~

      "
    `);
  });
});
