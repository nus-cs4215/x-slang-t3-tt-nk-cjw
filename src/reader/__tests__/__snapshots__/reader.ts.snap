// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`valid read tests basic atoms 1`] = `
Object {
  "_type": "SAtom",
  "val": "abc",
}
`;

exports[`valid read tests basic atoms 2`] = `
Object {
  "_type": "SAtom",
  "val": "abc->def",
}
`;

exports[`valid read tests basic booleans 1`] = `
Object {
  "_type": "SBoolean",
  "val": true,
}
`;

exports[`valid read tests basic booleans 2`] = `
Object {
  "_type": "SBoolean",
  "val": false,
}
`;

exports[`valid read tests basic cons 1`] = `
Object {
  "_type": "SList",
  "_variant": 0,
  "x": Object {
    "_type": "SAtom",
    "val": "a",
  },
  "y": Object {
    "_type": "SAtom",
    "val": "b",
  },
}
`;

exports[`valid read tests basic cons 2`] = `
Object {
  "_type": "SList",
  "_variant": 0,
  "x": Object {
    "_type": "SAtom",
    "val": "a",
  },
  "y": Object {
    "_type": "SList",
    "_variant": 0,
    "x": Object {
      "_type": "SAtom",
      "val": "b",
    },
    "y": Object {
      "_type": "SAtom",
      "val": "c",
    },
  },
}
`;

exports[`valid read tests basic lists 1`] = `
Object {
  "_type": "SNil",
}
`;

exports[`valid read tests basic lists 2`] = `
Object {
  "_type": "SList",
  "_variant": 0,
  "x": Object {
    "_type": "SAtom",
    "val": "abc",
  },
  "y": Object {
    "_type": "SNil",
  },
}
`;

exports[`valid read tests basic lists 3`] = `
Object {
  "_type": "SList",
  "_variant": 0,
  "x": Object {
    "_type": "SAtom",
    "val": "abc",
  },
  "y": Object {
    "_type": "SList",
    "_variant": 0,
    "x": Object {
      "_type": "SNumber",
      "val": 123,
    },
    "y": Object {
      "_type": "SNil",
    },
  },
}
`;

exports[`valid read tests basic lists 4`] = `
Object {
  "_type": "SList",
  "_variant": 0,
  "x": Object {
    "_type": "SAtom",
    "val": "abc",
  },
  "y": Object {
    "_type": "SList",
    "_variant": 0,
    "x": Object {
      "_type": "SNumber",
      "val": 123,
    },
    "y": Object {
      "_type": "SList",
      "_variant": 0,
      "x": Object {
        "_type": "SBoolean",
        "val": true,
      },
      "y": Object {
        "_type": "SNil",
      },
    },
  },
}
`;

exports[`valid read tests basic numbers 1`] = `
Object {
  "_type": "SNumber",
  "val": 123,
}
`;

exports[`valid read tests basic numbers 2`] = `
Object {
  "_type": "SNumber",
  "val": 123.5,
}
`;

exports[`valid read tests basic parens 1`] = `
Object {
  "_type": "SNil",
}
`;

exports[`valid read tests basic parens 2`] = `
Object {
  "_type": "SNil",
}
`;

exports[`valid read tests basic parens 3`] = `
Object {
  "_type": "SNil",
}
`;

exports[`valid read tests basic quotes 1`] = `
Object {
  "_type": "SList",
  "_variant": 0,
  "x": Object {
    "_type": "SAtom",
    "val": "quote",
  },
  "y": Object {
    "_type": "SList",
    "_variant": 0,
    "x": Object {
      "_type": "SNil",
    },
    "y": Object {
      "_type": "SNil",
    },
  },
}
`;

exports[`valid read tests basic quotes 2`] = `
Object {
  "_type": "SList",
  "_variant": 0,
  "x": Object {
    "_type": "SList",
    "_variant": 0,
    "x": Object {
      "_type": "SAtom",
      "val": "quote",
    },
    "y": Object {
      "_type": "SList",
      "_variant": 0,
      "x": Object {
        "_type": "SAtom",
        "val": "abc",
      },
      "y": Object {
        "_type": "SNil",
      },
    },
  },
  "y": Object {
    "_type": "SNil",
  },
}
`;

exports[`valid read tests basic quotes 3`] = `
Object {
  "_type": "SList",
  "_variant": 0,
  "x": Object {
    "_type": "SAtom",
    "val": "quote",
  },
  "y": Object {
    "_type": "SList",
    "_variant": 0,
    "x": Object {
      "_type": "SList",
      "_variant": 0,
      "x": Object {
        "_type": "SAtom",
        "val": "abc",
      },
      "y": Object {
        "_type": "SNil",
      },
    },
    "y": Object {
      "_type": "SNil",
    },
  },
}
`;

exports[`valid read tests basic quotes 4`] = `
Object {
  "_type": "SList",
  "_variant": 0,
  "x": Object {
    "_type": "SAtom",
    "val": "quote",
  },
  "y": Object {
    "_type": "SList",
    "_variant": 0,
    "x": Object {
      "_type": "SList",
      "_variant": 0,
      "x": Object {
        "_type": "SAtom",
        "val": "abc",
      },
      "y": Object {
        "_type": "SAtom",
        "val": "def",
      },
    },
    "y": Object {
      "_type": "SNil",
    },
  },
}
`;

exports[`valid read tests basic quotes 5`] = `
Object {
  "_type": "SList",
  "_variant": 0,
  "x": Object {
    "_type": "SAtom",
    "val": "quote",
  },
  "y": Object {
    "_type": "SList",
    "_variant": 0,
    "x": Object {
      "_type": "SList",
      "_variant": 0,
      "x": Object {
        "_type": "SAtom",
        "val": "quote",
      },
      "y": Object {
        "_type": "SList",
        "_variant": 0,
        "x": Object {
          "_type": "SList",
          "_variant": 0,
          "x": Object {
            "_type": "SAtom",
            "val": "abc",
          },
          "y": Object {
            "_type": "SNil",
          },
        },
        "y": Object {
          "_type": "SNil",
        },
      },
    },
    "y": Object {
      "_type": "SNil",
    },
  },
}
`;

exports[`valid read tests basic quotes 6`] = `
Object {
  "_type": "SList",
  "_variant": 0,
  "x": Object {
    "_type": "SAtom",
    "val": "quote",
  },
  "y": Object {
    "_type": "SList",
    "_variant": 0,
    "x": Object {
      "_type": "SList",
      "_variant": 0,
      "x": Object {
        "_type": "SList",
        "_variant": 0,
        "x": Object {
          "_type": "SAtom",
          "val": "quote",
        },
        "y": Object {
          "_type": "SList",
          "_variant": 0,
          "x": Object {
            "_type": "SAtom",
            "val": "abc",
          },
          "y": Object {
            "_type": "SNil",
          },
        },
      },
      "y": Object {
        "_type": "SNil",
      },
    },
    "y": Object {
      "_type": "SNil",
    },
  },
}
`;
